tensorflow_probability/python/experimental/distributions/laplace_approximation.py

```diff
@@ -0,0 +1,216 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""A Laplace approximation over a JointDistribution object."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow.compat.v2 as tf
+
+from tensorflow_probability.python.bijectors import reshape
+from tensorflow_probability.python.bijectors import split
+from tensorflow_probability.python.bijectors import chain
+from tensorflow_probability.python.bijectors import joint_map
+from tensorflow_probability.python.distributions import transformed_distribution
+from tensorflow_probability.python.distributions import mvn_tril
+from tensorflow_probability.python.math import value_and_gradient
+from tensorflow_probability.python.optimizer import bfgs_utils
+from tensorflow_probability.python.optimizer import bfgs
+
+
+def _common_dtype(dtypes, dtype_hint=None):
+  """The `tf.Dtype` common to all elements in `dtypes`."""
+  dtype = None
+  seen = []
+  for dt in dtypes:
+    seen.append(dt)
+    if dtype is None:
+      dtype = dt
+    elif dtype != dt:
+      raise TypeError(
+          "Found incompatible dtypes, {} and {}. Seen so far: {}".format(
+              dtype, dt, seen))
+  if dtype is None:
+    return dtype_hint
+  return dtype
+
+
+def _call(obj, fns, *args, **kwargs):
+  """attempt to call the first of `obj.fn` for `fn` in `fns`"""
+  for fn in fns:
+    call = getattr(obj, fn, None)
+    if call is not None:
+      return call(*args, **kwargs)
+  raise ValueError("Couldn't find any of {} in {}".format(fns, obj))
+
+
+def _sample(joint_dist, *args, **kwargs):
+  """Take samples regardless of joint distribution type."""
+  fns = ["sample", "sample_unpinned"]
+  return _call(joint_dist, fns, *args, **kwargs)
+
+
+def _log_prob(joint_dist, *args, **kwargs):
+  """Log prob regardless of joint distribution type."""
+  fns = ["log_prob", "unnormalized_log_prob"]
+  return _call(joint_dist, fns, *args, **kwargs)
+
+
+def _bfgs_dimension(joint_dist, bijector):
+  """The dimension which BFGS works on"""
+  flat_event_shape = tf.nest.flatten(joint_dist.event_shape)
+  return bijector.inverse_event_shape(flat_event_shape)[0]
+
+
+def _bfgs_minimize(
+    joint_dist,
+    bijector,
+    initial_values,
+    initial_inverse_hessian_estimate,
+    validate_convergence,
+):
+  """Minimize the -ve log prob of a joint distribution using BFGS."""
+
+  names = joint_dist._flat_resolve_names()
+
+  def loss(unconstrained_values):
+    constrained_values = bijector.forward(unconstrained_values)
+    return -_log_prob(joint_dist, **dict(zip(names, constrained_values)))
+
+  @tf.function(autograph=False)
```

Comment:
We try to avoid adding the function annotation internally, and let users add it where necessary. Would that be terrible here?

Reply:
yes I didn't want to add this, but you get an error about `None` gradients without it :( I wasn't sure if this is related to the same issue with gradients not working in jax and numpy modes (I don't have much understanding in that area). If you have any idea about how to fix that would be great. If you e.g. remove the annotation and debug one of the tests you can see:

``` python
(Pdb) loss_and_gradient(initial_position)
(<tf.Tensor: shape=(10,), dtype=float32, numpy=
array([4.7611666, 8.704665 , 4.0540056, 4.3684416, 3.7730975, 3.951736 ,
       4.1365232, 3.8006756, 4.6959515, 3.9844842], dtype=float32)>, None)
```